modules/gapi/src/executor/gexecutor.cpp

```diff
@@ -146,16 +146,23 @@ void writeBackExec(const Mag& mag, const RcDesc &rc, GRunArgP &g_arg)
         writeBack(mag, rc, g_arg);
         return;
     }
-    auto checkOutArgData = [&](const uchar* out_arg_data) {
-        //simply check that memory was not reallocated, i.e.
-        //both Mat and View pointing to the same memory
-        auto mag_data = mag.template slot<cv::RMat>().at(rc.id).get<RMatAdapter>()->data();
-        GAPI_Assert((out_arg_data == mag_data) && " data for output parameters was reallocated ?");
-    };
 
     switch (g_arg.index())
     {
-    case GRunArgP::index_of<cv::Mat*>() : checkOutArgData(util::get<cv::Mat*>(g_arg)->data); break;
+    case GRunArgP::index_of<cv::Mat*>() : {
+        // If there is a copy intrinsic at the end of the graph
+        // we need to actualy copy the data to the user buffer
+        // since output runarg was optimized to simply point
+        // to the input of the copy kernel
+        auto& out_mat = *util::get<cv::Mat*>(g_arg);
+        const auto& rmat = mag.template slot<cv::RMat>().at(rc.id);
+        auto mag_data = rmat.get<RMatAdapter>()->data();
+        if (out_mat.data != mag_data) {
+            auto view = rmat.access(RMat::Access::R);
+            asMat(view).copyTo(out_mat);
+        }
+        break;
+    }
```

Comment:
How all this worked previously:

1. GExecutor gets its output vector (`cv::GRunArgsP`), a vector of objects to write into
2. GExecutor binds those objects to the internal magazine using the internal identifiers
3. Islands in the script are running in order, every one is provided with input/output object pointers from a magazine
4. Finally, in write_back we check that outputs we has to produce and objects in magazine are the same. The removed check validated that a magazine-size object (`cv::Mat`) hasn't been reallocated along the way.

What we do now is just plain copy if pointers change.. I believe it is very /implicit/ way to handle copy-as-copy here...

Reply:
Added a FIXME